gateway/ln-gateway/src/state_machine/mod.rs

```diff
@@ -381,6 +383,52 @@ impl GatewayClientModule {
         }
     }
 
+    /// Attempts to remove a gateway's registration from the federation. Since
+    /// removing gateway registrations is best effort, this does not return
+    /// an error and simply emits a warning when the registration cannot be
+    /// removed.
+    pub async fn remove_from_federation(&self, gateway_keypair: KeyPair) {
+        // Removing gateway registrations is best effort, so just emit a warning if it
+        // fails
+        if let Err(e) = self.remove_from_federation_inner(gateway_keypair).await {
+            let gateway_id = gateway_keypair.public_key();
+            let federation_id = self.client_ctx.get_config().global.federation_id();
+            warn!("Failed to remove gateway {gateway_id} from federation {federation_id}: {e:?}");
+        }
+    }
+
+    /// Retrieves the signing challenge from each federation peer. Since each
+    /// peer maintains their own list of registered gateways, the gateway
+    /// needs to provide a signature that is signed by the private key of the
+    /// gateway id to remove the registration.
+    async fn remove_from_federation_inner(&self, gateway_keypair: KeyPair) -> anyhow::Result<()> {
+        let gateway_id = gateway_keypair.public_key();
+        let challenges = self
+            .module_api
+            .get_remove_gateway_challenge(gateway_id)
+            .await?;
+
+        let mut signatures: BTreeMap<PeerId, Signature> = BTreeMap::new();
+        for (peer_id, challenge) in challenges {
+            if let Some(challenge) = challenge {
+                let msg = Message::from_slice(&challenge)?;
```

Comment:
Security bug: We should not use the guardian-provided challenge directly, this would allow malicious guardians to make us sign anything. This isn't super bad since the gateway key isn't used for anything but registrations right now, but at least malicious guardians could de-register us from other federations.

We should build our actual signed message this way:

```
tag = "remove_request";
m = hash(tag + federation_id + guardian_id + challenge)
```

This way the message is only valid for the specific guardian and the use case of de-registering.

Reply:
Good catch! Fixed. I ended up using the Federation's public key instead of the federation id though, since it was easier to get inside the server module.